### PR TITLE
P5-205/calibration-solution

### PR DESCRIPTION
**Pull request by C0DK**
Rewrote the calibration section in solution. in case of typos etc, plz just fix em so it can be merged.
If you have any questions please write me on messenger.

### CODE REVIEW

**Comment by C0DK** [comment]
http://hcl.codes/teamcity/guestAuth/repository/download/P5_Build/.lastSuccessful/main.pdf?branch=P5-205/calibration-solution

File: report/5.Solution/MI_solution/calibration.tex [hunk 1]
```diff
@@ -1,7 +1,12 @@
-\section{Calibration}
+\section{Calibration}\label{sec:calibration}
+As explained in \autoref{des:sec:variablemotorpower} varying power was required depending on the current revolution of each motor, and the direction of movement.
 
-As the device required varying power in order to move, depending on the current revolution of the motor, the first goal was to extract these powers.
-In order to do so, a calibration function was implemented that will slowly increase the power on the motor, until it observes a change in revolution.
+To solve this problem, data was gathered for the minimum amount of power required to change the revolution.
+This data was then processed in a neural network, finding a relation between power and revolution in each direction.
+The neural network was then exported to C code and compiled on the NXT, and used as a base-value for motor power.
```
Comment (nillas12) [inline report/5.Solution/MI_solution/calibration.tex]: Er det ikke as C code

File: report/5.Solution/MI_solution/calibration.tex [hunk 2]
```diff
@@ -1,7 +1,12 @@
...
+
+To collect the data regarding the required power, a function was created on the NXT.
```
Comment (nillas12) [inline report/5.Solution/MI_solution/calibration.tex]: Ikke regarding

File: report/5.Solution/MI_solution/calibration.tex [hunk 3]
```diff
@@ -33,36 +38,69 @@ \section{Calibration}
 
 The \texttt{should\_stop\_moving} function will determine whether the motor has moved, or if the power exceeds the maximum allowed power, meaning it is unable to move further.
 The data is sent back to a host machine over the established USB connection.
-A plot of the data for moving the motor from the lowest possible revolution to the highest can be seen in \autoref{fig:calib_data}.
 
-\figur{1}{images/calib_data.png}{Required power for revolution on motor}{fig:calib_data}
+The function was run multiple times, and whenever the device was modified, the calibration function was run again.
+
+One of these data-sets, where the calibration process was run 5 times, can be seen in \autoref{fig:calibdata}.
```
Comment (nillas12) [inline report/5.Solution/MI_solution/calibration.tex]: Ikke one of these, hellere, et eksempel

File: report/5.Solution/MI_solution/calibration.tex [hunk 4]
```diff
@@ -33,36 +38,69 @@ \section{Calibration}
...
+This data-set is based on the y-axis, in an upwards direction.
+
+
+\figur{1}{images/calib_data.png}{Required power for revolution on motor}{fig:calibdata}
+
+There is a clear tendency in the data, even though it has some noise, which is presumably due to motor insecurities.
```
Comment (nillas12) [inline report/5.Solution/MI_solution/calibration.tex]: Ikke insecurities

File: report/5.Solution/MI_solution/calibration.tex [hunk 5]
```diff
@@ -33,36 +38,69 @@ \section{Calibration}
...
+The tendency seems reasonable regarding the upwards motion, where it initially requires low effort, but the requirement increases as it is reaching horizontal level.
```
Comment (nillas12) [inline report/5.Solution/MI_solution/calibration.tex]: Ikke reaching eller whatever

File: report/5.Solution/MI_solution/calibration.tex [hunk 6]
```diff
@@ -33,36 +38,69 @@ \section{Calibration}
...
+
+Using this data, a neural network was trained to find the tendency of the data.
+The training was done with the Python module, scikit-learn\cite{scikit-learn}, which exposes a regressor called \texttt{MPLRegressor}, which was utilized as shown in \autoref{lst:mlpregressor}.
+A custom modified version of the module was created, to be able to use a custom activation function.
+This was done because of issues with the nxtOSEK system, regarding the usage of the \texttt{math.h} library in C, which was not usable for among others the \texttt{exp} function.
```
Comment (nillas12) [inline report/5.Solution/MI_solution/calibration.tex]: Den her sætning

File: report/5.Solution/MI_solution/calibration.tex [hunk 7]
```diff
@@ -33,36 +38,69 @@ \section{Calibration}
...
+model.fit(inp, expect)
 
-regressor.fit(input_values, expected_values)
 \end{lstlisting}
 
-The \texttt{MLPRegressor} has a lot of parameters that can be modified in order to tweak the neural network for your use case.
-In this case default values are used for everything except solver, activation and size of hidden layers.
-The activation function is set to use the logistic sigmoid function, and the model is defined to have 2 hidden layers with 30 neurons in each.
-These values are obtained by trying out a lot of different values, in order to determine which values generally perform well on the data set.
+The \texttt{MLPRegressor} has a lot of parameters that can be modified in order to tweak the neural network for the required use case.
```
Comment (nillas12) [inline report/5.Solution/MI_solution/calibration.tex]: Lidt ligegyldig sætning

File: report/5.Solution/MI_solution/calibration.tex [hunk 8]
```diff
@@ -33,36 +38,69 @@ \section{Calibration}
...
+A custom amount of hidden layers was required as the default argument value was 100 neurons in one layer, which was not necessary for the rather simple data.
+30 neurons in one layer was decided upon after testing different sizes.
+A larger hidden layer did not improve the quality, however smaller networks seemed to decrease precision.
```
Comment (nillas12) [inline report/5.Solution/MI_solution/calibration.tex]: Der bliver repeat i de her sætninger

File: report/5.Solution/MI_solution/calibration.tex [hunk 9]
```diff
@@ -33,36 +38,69 @@ \section{Calibration}
...
+
+The system also utilized the custom implementation of the sigmoid function, and increased max iterations, and lowered the tolerance to make the network run for a longer time before stopping.
```
Comment (nillas12) [inline report/5.Solution/MI_solution/calibration.tex]: Ved godt vi gør det, den her sætning er bare lidt nævn, nævn, nævn

File: report/5.Solution/MI_solution/calibration.tex [hunk 10]
```diff
@@ -33,36 +38,69 @@ \section{Calibration}
...
+This makes sure the network is presumably good enough to be used in the system.
```
Comment (nillas12) [inline report/5.Solution/MI_solution/calibration.tex]: Ensures, fjern presume

File: report/5.Solution/MI_solution/calibration.tex [hunk 11]
```diff
@@ -33,36 +38,69 @@ \section{Calibration}
...
 
-\figur{1}{images/calib_data_nn.png}{Sample fitting of the data}{fig:calib_data}
+\figur{1}{images/calib_data_nn.png}{Sample fitting of the data}{fig:calibdatann}
 
-Now this model has to be exported to the NXT. 
-In order to do so, some C code is generated based on the weights and biases from the model.
-The C code contains all the weights and biases along with the activation function and a function to calculate the output given input.
+\autoref{fig:calibdatann} shows the neural network for the upwards direction, however the same process was done simultaneously for the downwards direction.
+Calibration was only done for the $Y$-axis.
+The two separate neural networks, produced at another calibration test, can be seen in \autoref{fig:calibdatannboth}.
```
Comment (nillas12) [inline report/5.Solution/MI_solution/calibration.tex]: On nok for also produced

File: report/5.Solution/MI_solution/calibration.tex [hunk 12]
```diff
@@ -33,36 +38,69 @@ \section{Calibration}
...
 
-The generated model is big, but a simplified, condensed version can be seen in \autoref{lst:exported_model}:
 
+\figur{1}{images/calib_data_nn_both.png}{Both neural networks of the $Y$ axis}{fig:calibdatannboth}
+
+These neural networks had to be used to calculate the base value of a motion at a given revolution at a given time.
+To solve this, the weight and biases of the neural networks had to be exported to the NXT device.
+This was done by generating the C code needed to replicate the behavior.
+A simplistic version of such a file can be seen in \autoref{lst:exportedmodel}.
```
Comment (nillas12) [inline report/5.Solution/MI_solution/calibration.tex]: Ikke file

File: report/5.Solution/MI_solution/calibration.tex [hunk 13]
```diff
@@ -33,36 +38,69 @@ \section{Calibration}
...
+Calibration was only done for the $Y$-axis.
+The two separate neural networks, produced at another calibration test, can be seen in \autoref{fig:calibdatannboth}.
 
-The generated model is big, but a simplified, condensed version can be seen in \autoref{lst:exported_model}:
 
+\figur{1}{images/calib_data_nn_both.png}{Both neural networks of the $Y$ axis}{fig:calibdatannboth}
+
+These neural networks had to be used to calculate the base value of a motion at a given revolution at a given time.
+To solve this, the weight and biases of the neural networks had to be exported to the NXT device.
+This was done by generating the C code needed to replicate the behavior.
```
Comment (nillas12) [inline report/5.Solution/MI_solution/calibration.tex]: Behøver vi ikke, kan slås sammen med nedenstående

File: report/5.Solution/Scheduling.tex [hunk 14]
```diff
@@ -76,7 +76,7 @@ \section{Scheduling}\label{solution:scheduling}
 };
 \end{lstlisting}
 
-The reasoning behind simply having as major cycle with no minor cycles is the execution time of the individual methods, which can be seen on \autoref{table:executionTimes}.
+The reasoning behind simply having as major cycle with no minor cycles is the execution time of the individual methods, which can be seen on \autoref{table:executionTimes1}.
```
Comment (nillas12) [inline report/5.Solution/Scheduling.tex]: As major cycle

File: report/5.Solution/Calibration.tex [hunk 15]
```diff
@@ -1 +1 @@
-\section{Calibration}\label{sec:calibration}
\ No newline at end of file
+\section{Calibration}\label{}
```
Comment (frederikspang) [inline report/5.Solution/Calibration.tex]: Hvorfor tom label?

File: report/5.Solution/MI_solution/calibration.tex [hunk 16]
```diff
@@ -33,36 +38,69 @@ \section{Calibration}
 
 The \texttt{should\_stop\_moving} function will determine whether the motor has moved, or if the power exceeds the maximum allowed power, meaning it is unable to move further.
 The data is sent back to a host machine over the established USB connection.
-A plot of the data for moving the motor from the lowest possible revolution to the highest can be seen in \autoref{fig:calib_data}.
 
-\figur{1}{images/calib_data.png}{Required power for revolution on motor}{fig:calib_data}
+The function was run multiple times, and whenever the device was modified, the calibration function was run again.
+
+One of these data-sets, where the calibration process was run 5 times, can be seen in \autoref{fig:calibdata}.
+This data-set is based on the y-axis, in an upwards direction.
```
Comment (frederikspang) [inline report/5.Solution/MI_solution/calibration.tex]:
```suggestion
This data-set is based on the y-axis in an upwards direction.
```

File: report/5.Solution/MI_solution/calibration.tex [hunk 17]
```diff
@@ -33,36 +38,69 @@ \section{Calibration}
...
+\figur{1}{images/calib_data.png}{Required power for revolution on motor}{fig:calibdata}
+
+There is a clear tendency in the data, even though it has some noise, which is presumably due to motor insecurities.
+The tendency seems reasonable regarding the upwards motion, where it initially requires low effort, but the requirement increases as it is reaching horizontal level.
+
+Using this data, a neural network was trained to find the tendency of the data.
+The training was done with the Python module, scikit-learn\cite{scikit-learn}, which exposes a regressor called \texttt{MPLRegressor}, which was utilized as shown in \autoref{lst:mlpregressor}.
+A custom modified version of the module was created, to be able to use a custom activation function.
+This was done because of issues with the nxtOSEK system, regarding the usage of the \texttt{math.h} library in C, which was not usable for among others the \texttt{exp} function.
+Because of this an approximation of the sigmoid function was implemented.
```
Comment (frederikspang) [inline report/5.Solution/MI_solution/calibration.tex]:
```suggestion
Because of this an approximation of the \texttt{sigmoid} function was implemented.
```

File: report/5.Solution/Calibration.tex (hunk 15 above)
Comment (C0DK) [reply report/5.Solution/Calibration.tex]: tom fil. den skal bare slettes
